### PR TITLE
Optimize _bulk_get endpoint

### PR DESCRIPTION
**Pull request by nickva**
Use the new `fabric:open_revs/3` API implemented in #4201 to optimize the `_bulk_get` HTTP API. Since `open_revs/3` itself is new, allow reverting to individual doc fetches using the previous `open_revs/4` API via a config setting, mostly as a precautionary measure.

The implementation consists of three main parts:
  * Parse and validate args
  * Fetch the docs using `open_revs/3` or `open_revs/4`
  * Emit results as json or multipart, based on the `Accept` header value

Parsing and validation checks for various errors and then returns a map of `#{Ref => {DocId, RevOrError, DocOptions}}` and a list of Refs in the original argument order. The middle tuple element of `RevOrError` is notable that it may hold either the revision ID (`[Rev]` or `all`) or `{error, {Rev, ErrorTag, ErrorReason}}`.

Fetching the docs is fairly straightforward. The slightly interesting aspect is when an error is returned from `open_revs/3` we have to pretend that all the batched docs failed with that error. That is done to preserve the "zip" property, where all the input arguments have their matching result at the same position in results list. Another notable thing here is we fixed a bug where the error returned from `fabric:open_revs/3,4` was not formatted in a way it could have been emitted as json resulting in a function clause. That is why we call `couch_util:to_binary/1` on it. This was detected by the integration testing outline before and was missed by the previous mocked unit test.

The last part is emitting the results as either json or multipart. Here most changes are cleanups and grouping into separate handler functions. The `Accept` header can be either `multipart/related` or `multipart/mixed` and we try to emit the same content type as it was passed in the `Accept` header. One notable thing here is by DRY-ing the filtering of attachments in `non_stubbed_attachments/1` we fixed another bug when the multipart result was returning nonsense in cases when all attachments were stubs. The doc was returned as a multipart chunk with content type `multipart/...` instead of application/json. This was also detected in the integration tests described below.

The largest changes are in the testing area. Previous multipart tests were using mocks heavily, were quite fragile, and didn't have good coverage. Those tests were removed and replaced by new end-to-end tests in `chttpd_bulk_get_test.erl`. To make that happen add a simple multipart parser utility function which knows how to parse multipart responses into maps. Those maps preserve chunk headers and we can match those with `?assertMatch(...)` fairly easily. The tests try to get decent coverage for `chttpd_db.erl` bulk_get implementation and its utility functions, but they are also end-to-end tests so they test everything below, including fabric and couch layers as well.

Quick 1 node testing using the couchdyno replicating of 1 million docs shows at least a 2x speedup to complete the replication using this PR.

On main:

```
r=rep.Rep(); r.replicate_1_to_n_and_compare(1, num=1000000, normal=True)
330 sec
```

With this PR:
```
r=rep.Rep(); r.replicate_1_to_n_and_compare(1, num=1000000, normal=True)
160 sec
```

Individual `_bulk_get` response times shows an even higher improvement: an 8x speedup:

On main:
```
[notice] ... POST /cdyno-0000001/_bulk_get?latest=true&revs=true&attachments=false 200 ok 468
[notice] ... POST /cdyno-0000001/_bulk_get?latest=true&revs=true&attachments=false 200 ok 479
```

With this PR:
```
[notice] ... POST /cdyno-0000001/_bulk_get?latest=true&revs=true&attachments=false 200 ok 54
[notice] ... POST /cdyno-0000001/_bulk_get?latest=true&revs=true&attachments=false 200 ok 61
```

Fixes: https://github.com/apache/couchdb/issues/4183

### CODE REVIEW

**Comment by nickva** [comment]
This PR also fixed https://github.com/apache/couchdb/issues/3328

File: src/chttpd/test/eunit/chttpd_bulk_get_test.erl [hunk 1]
```diff
@@ -654,6 +1303,34 @@ t_missing_rev_latest({_, DbUrl}) ->
         Res
     ).
 
+t_fabric_worker_error({_, DbUrl}) ->
+    % Check the case of latest and a missing rev
```
Comment (jaydoane) [inline src/chttpd/test/eunit/chttpd_bulk_get_test.erl:1307]: Copy-pasta error?

**Review by jaydoane — APPROVED** [review]
great work, and amazing performance improvement 🚀 
```
module 'chttpd_bulk_get_test'
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_invalid_method)...[0.004 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_empty_request)...[0.019 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_no_docs)...[0.004 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_invalid_query_params)...[0.018 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_invalid_doc)...[0.004 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_doc_no_id)...ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_invalid_doc_id)...ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_missing_doc)...ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_invalid_rev)...ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_missing_rev)...ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_doc_all_revs)...[0.001 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_specific_rev)...[0.001 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_specific_rev_latest)...[0.001 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_ancestor_rev_latest)...[0.001 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_revs_true)...[0.001 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_attachments_true)...[0.001 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_atts_since)...ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_invalid_atts_since)...ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_invalid_atts_since_invalid_rev)...ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_atts_since_returns_attachment)...[0.001 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_atts_since_overrides_attachments_true)...ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_atts_since_multiple)...[0.001 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_atts_since_multiple_attachments_true)...[0.001 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_missing_rev_latest)...[0.001 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_fabric_worker_error)...ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_mp_empty_request)...[0.005 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_mp_no_docs)...[0.003 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_mp_invalid_doc)...[0.003 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_mp_doc_no_id)...[0.003 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_mp_invalid_doc_id)...[0.004 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_mp_missing_doc)...[0.003 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_mp_invalid_rev)...[0.004 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_mp_missing_rev)...[0.004 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_mp_doc_all_revs)...[0.025 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_mp_specific_rev)...[0.001 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_mp_specific_rev_multipart_related)...[0.001 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_mp_revs_true)...[0.001 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_mp_atts_since)...ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_mp_atts_since_returns_attachment)...[0.001 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_mp_atts_since_overrides_attachments_true)...ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_invalid_method)...[0.010 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_empty_request)...[0.005 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_no_docs)...[0.011 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_invalid_query_params)...[0.018 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_invalid_doc)...[0.010 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_doc_no_id)...ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_invalid_doc_id)...ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_missing_doc)...[0.012 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_invalid_rev)...ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_missing_rev)...[0.001 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_doc_all_revs)...ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_specific_rev)...ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_specific_rev_latest)...[0.001 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_ancestor_rev_latest)...[0.002 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_revs_true)...[0.001 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_attachments_true)...[0.008 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_atts_since)...[0.001 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_invalid_atts_since)...ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_invalid_atts_since_invalid_rev)...ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_atts_since_returns_attachment)...ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_atts_since_overrides_attachments_true)...ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_atts_since_multiple)...[0.001 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_atts_since_multiple_attachments_true)...[0.001 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_missing_rev_latest)...ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_fabric_worker_error)...ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_mp_empty_request)...[0.009 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_mp_no_docs)...[0.004 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_mp_invalid_doc)...[0.015 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_mp_doc_no_id)...[0.005 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_mp_invalid_doc_id)...[0.004 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_mp_missing_doc)...[0.004 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_mp_invalid_rev)...[0.004 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_mp_missing_rev)...[0.009 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_mp_doc_all_revs)...[0.001 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_mp_specific_rev)...[0.001 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_mp_specific_rev_multipart_related)...[0.001 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_mp_revs_true)...[0.001 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_mp_atts_since)...[0.001 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_mp_atts_since_returns_attachment)...[0.001 s] ok
  chttpd_bulk_get_test:97: -with/1-fun-0- (t_mp_atts_since_overrides_attachments_true)...[0.001 s] ok
  chttpd_bulk_get_test:125: -bulk_get_multiple_docs_test_/0-fun-2- (t_multiple_docs)...[0.107 s] ok
  chttpd_bulk_get_test:126: -bulk_get_multiple_docs_test_/0-fun-0- (t_mp_multiple_docs)...[0.175 s] ok
  chttpd_bulk_get_test:136: -bulk_get_multiple_docs_no_batches_test_/0-fun-2- (t_multiple_docs)...[0.443 s] ok
  chttpd_bulk_get_test:137: -bulk_get_multiple_docs_no_batches_test_/0-fun-0- (t_mp_multiple_docs)...[0.779 s] ok
  [done in 5.117 s]
```

File: src/chttpd/src/chttpd_db.erl [hunk 2]
```diff
@@ -2363,6 +2288,175 @@ set_namespace(NS, #mrargs{} = Args) ->
...
+        case Response of
+            [] ->
+                % Remap empty reponses to errors. This is a peculiarity of the
+                % _bulk_get HTTP API. If revision was not specifed, `undefined`
+                % must be returned as the error revision ID.
+                #{Ref := {_, Revs, _}} = ArgsMap,
+                RevStr = bulk_get_rev_error(Revs),
+                Error = {RevStr, <<"not_found">>, <<"missing">>},
+                {Ref, {DocId, {error, Error}, []}};
+            [_ | _] = DocRevisions ->
```
[the remark this answers is not in the document]
Comment (nickva) [reply src/chttpd/src/chttpd_db.erl:2328]: [Yup](https://medium.com/erlang-battleground/ode-to-the-robot-butt-bbd69e69beb2)!

File: src/chttpd/test/eunit/chttpd_bulk_get_test.erl (hunk 1 above)
Comment (nickva) [reply src/chttpd/test/eunit/chttpd_bulk_get_test.erl]: Yes it is. Good find!

**Pull request merged**